### DISCUSSION TITLE
Multiplayer spectate does not advance on maps with videos

### DISCUSSION THREAD

**Discussion by muffyn** [ppy/osu#15825] [Q&A] [answered]
Storyboards and normal maps work fine. Maps with video cause spectate to just pause at the spot the spectate started.

https://streamable.com/thmhpj

2021.1120.0-lazer

[database.log](https://github.com/ppy/osu/files/7610247/database.log)
[network.log](https://github.com/ppy/osu/files/7610248/network.log)
[performance.log](https://github.com/ppy/osu/files/7610249/performance.log)
[runtime.log](https://github.com/ppy/osu/files/7610250/runtime.log)
[updater.log](https://github.com/ppy/osu/files/7610251/updater.log)

2021-11-26 18:07:31 [verbose]: Video too far out of sync (41.70837504170837), seeking to 4851.2883193517
2021-11-26 18:07:31 [verbose]: Video too far out of sync (208.54187520854185), seeking to 4851.2883193517

**Comment by bdach** [comment]
i've verified this and as far as i can tell the assertion that this happens on all "maps with videos" *not* correct. it looks to be just this one specific map.

* https://osu.ppy.sh/beatmapsets/873639#osu/1826084 works correctly
* https://osu.ppy.sh/beatmapsets/336822#osu/745345 (depicted on video above) does not work correctly.

@Muffyn please confirm that this is the case on your end too before i escalate this to an issue, to confirm that we are seeing the same thing.

**Reply by muffyn** [reply]
The KDA map works correctly for me too. I think it was just a coincidence that it happened on two video maps in a row. I managed to reproduce it accidentally with a non-video map too.

Video, ranked today:
https://osu.ppy.sh/beatmapsets/1119169#osu/2340750

Minute long video showing that this map does not work, but other maps do work. The next map selected in the room worked, too.
https://streamable.com/fheify
Bug stats happening from 2021-11-26 17:41:43 [verbose]: Game-wide working beatmap updated to Yorushika - Itte. (Xen) [Insane]
Started recording at 2021-11-26 17:43:59
[database.log](https://github.com/ppy/osu/files/7610637/database.log)
[network.log](https://github.com/ppy/osu/files/7610638/network.log)
[performance.log](https://github.com/ppy/osu/files/7610639/performance.log)
[runtime.log](https://github.com/ppy/osu/files/7610640/runtime.log)
[updater.log](https://github.com/ppy/osu/files/7610641/updater.log)

No video, old:
https://streamable.com/ckbgyo
2021-11-26 20:41:05 [verbose]: Game-wide working beatmap updated to Yuyoyuppe - AiAe (Fort) [Eternal]
Entered spectator twice, paused both times:
2021-11-26 20:41:19 [verbose]: Screen changed ← Multiplayer
2021-11-26 20:41:20 [verbose]: Screen changed → MultiSpectatorScreen
2021-11-26 20:41:40 [verbose]: Screen changed ← Multiplayer
2021-11-26 20:41:41 [verbose]: Screen changed → MultiSpectatorScreen
2021-11-26 20:41:49 [verbose]: Screen changed ← Multiplayer
https://osu.ppy.sh/beatmapsets/257793#osu/589969
[database.log](https://github.com/ppy/osu/files/7610630/database.log)
[network.log](https://github.com/ppy/osu/files/7610631/network.log)
[performance.log](https://github.com/ppy/osu/files/7610632/performance.log)
[runtime.log](https://github.com/ppy/osu/files/7610633/runtime.log)
[updater.log](https://github.com/ppy/osu/files/7610634/updater.log)

Hope this helps investigate.

**Reply by peppy** [reply]
If possible could you identify in the logs what timestamp the issue occurred? There's quite a bit of content in the logs which doesn't seem related, would be very helpful.

**Reply by muffyn** [reply]
I've updated my previous reply with timestamps + videos (recordings start late on both)

**Comment by frenzibyte** [comment, marked as answer]
This very likely goes back to the old https://github.com/ppy/osu/issues/14162, where some beatmaps that affect the gameplay start time in some way (via either a storyboard element like a video, or `AudioLeadIn`) in combination with user/platform clock offset causing the multiplayer spectator to freeze or starting running backwards.

**Reply by frenzibyte** [reply]
Can confirm that is the case, all of the three beatmaps mentioned above (one decipted from video and two mentioned in reply) have a negative [`startOffset`](https://github.com/ppy/osu/blob/5f27f1c099c0b1e369cf01662bcf758de91ae42d/osu.Game/Screens/Play/MasterGameplayClockContainer.cs#L77-L96) value:
 - https://osu.ppy.sh/beatmapsets/336822#osu/745345: -1230ms, due to first hitobject starting at 770ms.
 - https://osu.ppy.sh/beatmapsets/1119169#osu/2340750: -925ms, due to earliest storyboard element at -925ms.
 - https://osu.ppy.sh/beatmapsets/257793#osu/589969: -992ms, due to first hitobject starting at 1008ms.

Any beatmap with a `startOffset <= -28` could end up freezing the spectator.

**Reply by peppy** [reply]
@smoogipoo worth checking on the above while you're looking into a very similar issues (curious if you encountered this as well?)

**Reply by frenzibyte** [reply]
FWIW I have a failing test case for that as well: 4f4f60248faad019bc3c0501e42de854fea5dbdc.

I also have a potential solution but I'm still gathering all the pieces from my previous investigations to make sense of what's going on first before proceeding to PR (I can say my previous attempts at explaining this were indeed hard to follow...)

**Reply by peppy** [reply]
Was the potential solution something that was addressed in your recent PRs? Or is this still an issue?

**Reply by frenzibyte** [reply]
Yeah it should be resolved by https://github.com/ppy/osu/pull/16713 which includes the test case mentioned above.
